experimental/gcp-log/function.go

```diff
@@ -156,37 +210,49 @@ func Sequence(w http.ResponseWriter, r *http.Request) {
 
 // Integrate is the entrypoint of the `integrate` GCF function.
 func Integrate(w http.ResponseWriter, r *http.Request) {
-	var d struct {
-		Origin     string `json:"origin"`
-		Initialise bool   `json:"initialise"`
-		Bucket     string `json:"bucket"`
-	}
+	// process request args
 
+	d := requestData{}
 	if err := json.NewDecoder(r.Body).Decode(&d); err != nil {
 		fmt.Printf("json.NewDecoder: %v", err)
-		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+		http.Error(w, fmt.Sprintf("Failed to decode JSON: %q", err), http.StatusBadRequest)
 		return
 	}
 
-	ok, pubKey := validateCommonArgs(w, d.Origin)
-	if !ok {
+	if ok := validateCommonArgs(w, d); !ok {
 		return
 	}
 
-	privKey := os.Getenv("SERVERLESS_LOG_PRIVATE_KEY")
-	if len(privKey) == 0 {
-		http.Error(w,
-			"Please set SERVERLESS_LOG_PUBLIC_KEY environment variable",
-			http.StatusBadRequest)
+	kmsKeyName := fmt.Sprintf(kmssigner.KeyVersionNameFormat,
```

Comment:
Hmm, is it worth factoring out the L226-L254 in this method and similar block in `Sequence`?